doc/dev/CustomRepoGuide.md

```diff
@@ -0,0 +1,387 @@
+~~~
+
+The `if [ [ "$COVERAGE" == "--coverage" ] ]; then ... ;fi` part of the code tells travis to only execute these lines if the environmental variable `$COVERAGE` is set to "--coverage". This has nothing to do with codecov itself. It just prevents your travis build from failing when no codecov data is generated. If you have a look at the CMake call in the `.travis.yml`, you will find this flag: `-DCMAKE_CXX_FLAGS="$COVERAGE $WERROR"`. So the variable is also passed to Cmake. CMake itself will pass it to the compiler. The flag "--coverage" tells the compiler to generate coverage data. Without it, there will be no data to process and the codecov commands will fail (and consequently your tests fail too). Hence we use the if block. 
+Let's have a look at the actual command passed to the docker image: 
+
+~~~bash
+lcov --directory . --capture --output-file ../source/coverage.info
+~~~
+
+The program lcov just processes the coverage data generated by the programs. It is important to know, that you actually need to run your programs to generate the data. Compiling them is not enough! `--directory . --capture` tells lcov to gather the data in the current directory. '--output-file ../source/coverage.info' tells lcov to write the result file into your source directory (inside the docker image we are in "/home/nuto/build"). Now here are two possible pitfalls:
+First we need root access inside the image as stated in the travis section above. This is done by starting the image with `-u 0`. The reason is, that our NuTo sources are just mounted to the directory "/home/nuto/source" and in our current image you don't have write permission to that directory when starting as the default user.
+The second proble is, that you need to upload the coverage report from the source directory and not the build or any other directory. Otherwise the output from the codecov website might not give you the expected results.
+
+The second command
+
+~~~bash
+bash <(curl -s https://codecov.io/bash)
+~~~
+
+does the actual upload to the codecov website. Notice that we don't use the docker image here! Since we have written the data to the shared directory it is not necessary! And thats basically it!
+
+**Additional notes**
+
+
+If you don't want certain folders to be included into your coverage report you can just add the following line between the other two commands:
+
+~~~yml
+- if [ [ "$COVERAGE" == "--coverage" ] ]; then docker exec dock lcov --remove ../source/coverage.info '${FOLDER_TO_EXCLUDE}' --output-file ../source/coverage.info; fi
+~~~
+
+Typical folders might be "usr/*" to exclude system libraries or "*/test*" to exclude your testfiles. The latter one won't remove the code from everything that is tested, just the code from your actual testfiles itself. Usually you are not interested of how much files in your testfile are run, just how many lines of your tested objects are hit.
```

Comment:
I would disagree with the last sentence. Maybe you're not running all the tests you are think you are running. Coverage on the tests themselves would reveal that.

Reply:
True to some degree. I think I just remove this part, since it contains no additional information and might cause some confusion.